### PR TITLE
Michael#1/authentication functionality

### LINKED ISSUE #1: Setup Sign up and Login Functionalities with rails
-  Add login and sign up  functionalities with rails backend
-  Each user should have a username , email and password

### PR DESCRIPTION
**Pull request by MICHAELMUNAVU83**
- Added users with a username , email and password
- Added login and signup logic for users
- Added validations to ensure valid users sign up

### CODE REVIEW

File: api/app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,2 +1,41 @@
 class ApplicationController < ActionController::API
+    before_action :authorized
+    def encode_token(payload)
+        JWT.encode(payload, 'my_s3cr3t')
```
Comment (otsembo) [inline api/app/controllers/application_controller.rb:4]: Remove the JWT secret in the code. Alternatively, add it as an environment variable.

File: api/app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+    has_secure_password
+    validates :email, uniqueness: true
+    validates :email, presence: true
```
Comment (otsembo) [inline api/app/models/user.rb]: You can combine multiple validations in the same line as opposed to making a repeat.
Eg `validates :email, { uniqueness: true, presence: true }`

**Review by otsembo — APPROVED** [review]
Hello @MICHAELMUNAVU83, Good job on the PR
Check on the few fixes required and then merge the code.

File: api/app/controllers/application_controller.rb (hunk 1 above)
Comment (MICHAELMUNAVU83) [reply api/app/controllers/application_controller.rb:4]: Hello @otsembo  , Could you clarify on how I can add it as an environment variable?

Comment (Muriithi-Gitonga) [reply api/app/controllers/application_controller.rb:4]: Hello @MICHAELMUNAVU83  I can help you with this

Comment (MICHAELMUNAVU83) [reply api/app/controllers/application_controller.rb:4]: Thanks @Muriithi-Gitonga , how would I go about it

Comment (Muriithi-Gitonga) [reply api/app/controllers/application_controller.rb:4]: *They are many approaches to do this, for me I used the credentials.yml:*
You open the file through this command from your terminal `EDITOR="code --wait" bin/rails credentials:edit` you can change the editor type if you are not using vscode. When the file opens enter your secret_key as key:value pair `jwt_secret: secret` then you can access the secret key through `Rails.application.credentials.jwt_secret` ***You can check this [video](https://www.youtube.com/watch?v=vJco2_0Y7hY&t=1062s)***

**Pull request merged**